json-rpc/src/main/scala/com/github/apalachemc/apalache/jsonrpc/JsonRpcServer.scala

```diff
@@ -165,67 +170,68 @@ class ExplorationService(config: Try[Config.ApalacheConfig]) extends LazyLogging
     // Prepare the transition. Make sure that we do not update the SMT context concurrently.
     validationResult
       .map { checkerContext =>
-        {
-          checkerContext.synchronized {
-            var snapshotBeforeId: Integer = -1
-            if (params.rollbackToSnapshotId >= 0) {
-              // try to recover the snapshot
-              val recovered = snapshots.recoverSnapshot(sessionId, checkerContext, params.rollbackToSnapshotId)
-              if (!recovered) {
-                return Left(ServiceError(JsonRpcCodes.INVALID_PARAMS,
-                        s"Snapshot not found: ${params.rollbackToSnapshotId} in session $sessionId"))
-              }
-              snapshotBeforeId = params.rollbackToSnapshotId
-            } else {
-              // take a snapshot of the current context
-              snapshotBeforeId = snapshots.takeSnapshot(sessionId, checkerContext)
-            }
-            // the step number might have changed after recovery, so we re-fetch the transitions
-            val stepNo = checkerContext.trex.stepNo
-            val transitions = if (stepNo == 0) {
-              checkerContext.checkerInput.initTransitions
-            } else {
-              checkerContext.checkerInput.nextTransitions
-            }
-            if (transitionId < 0 || transitionId >= transitions.size) {
+        withLock(params.sessionId) {
+          if (!sessions.contains(params.sessionId)) {
+            return Left(ServiceError(JsonRpcCodes.INVALID_PARAMS, s"Session not found: ${params.sessionId}"))
```

Comment:
wouldn't this fail earlier – in `withLock` while trying to retrieve the lock?

(pattern appears a couple of times below)